packages/providers/src/transaction-summary/operations.ts

```diff
@@ -128,43 +139,37 @@ function isSameOperation(a: Operation, b: Operation) {
 
 /** @hidden */
 export function addOperation(operations: Operation[], toAdd: Operation) {
-  const ops = operations
-    .map((op) => {
-      // if it's not same operation, don't change. we just wanna stack the same operation
-      if (!isSameOperation(op, toAdd)) {
-        return null;
-      }
-
-      let newOp = { ...op };
-
-      // if it's adding new assets
-      if (toAdd.assetsSent?.length) {
-        // if prev op had assets, merge them. Otherwise just add the new assets
-        newOp = {
-          ...newOp,
-          assetsSent: op.assetsSent?.length ? mergeAssets(op, toAdd) : toAdd.assetsSent,
-        };
-      }
+  const allOperations = [...operations];
+
+  // Verifying if the operation to add already exists.
+  const index = allOperations.findIndex((op) => isSameOperation(op, toAdd));
+
+  if (allOperations[index]) {
+    // Existent operation, we want to edit it.
+    const existentOperation = { ...allOperations[index] };
+
+    if (toAdd.assetsSent?.length) {
+      /**
+       * If the assetSent already exists, we call 'mergeAssets' to merge possible
+       * entries of the same 'assetId', otherwise we just add the new 'assetSent'.
+       */
+      existentOperation.assetsSent = existentOperation.assetsSent?.length
+        ? mergeAssets(existentOperation, toAdd)
+        : toAdd.assetsSent;
+    }
 
-      // if it's adding new calls,
-      if (toAdd.calls?.length) {
-        /*
-[]          for calls we don't stack as grouping is not desired.
-          we wanna show all calls in the same operation
-          with each respective assets, amounts, functions, arguments.
-        */
-        newOp = {
-          ...newOp,
-          calls: [...(op.calls || []), ...(toAdd.calls || [])],
-        };
-      }
+    if (toAdd.calls?.length) {
+      // We need to stack the new call(s) with the possible existent ones.
+      existentOperation.calls = [...(existentOperation.calls || []), ...toAdd.calls];
+    }
 
-      return newOp;
-    })
-    .filter(Boolean) as Operation[];
+    allOperations[index] = existentOperation;
+  } else {
+    // New operation, we can simply add it.
+    allOperations.push(toAdd);
+  }
 
-  // if this operation didn't exist before just add it to the end
-  return ops.length ? ops : [...operations, toAdd];
+  return allOperations;
 }
```

Comment:
Here's another take on it. After looking into the code, IMO the comments are redundant as they're just explaining control flow and aren't giving any additional context.

```suggestion
/** @hidden */
export function addOperation(operations: Operation[], toAdd: Operation) {
  const index = operations.findIndex((op) => isSameOperation(op, toAdd));

  if (index === -1) {
    return [...operations, toAdd];
  }

  const operation = operations[index];

  const updatedOperation = {
    ...operation,
    assetsSent: mergeAssets(operation, toAdd),
    calls: [...(operation.calls || []), ...(toAdd.calls ?? [])],
  };

  return [
    ...operations.slice(0, index),
    updatedOperation,
    ...operations.slice(index + 1, operations.length),
  ];
}
```